components/server/src/api/server.ts

```diff
@@ -50,4 +73,120 @@ export class API {
+                        router.service(type, new Proxy(impl, this.interceptService(type)));
+                    }
+                },
+            }),
+        );
+        // TODO(al) cover unhandled cases
+    }
+
+    /**
+     * intercept handles cross-cutting concerns for all calls:
+     * - authentication
+     * - server-side observability
+     * TODO(ak):
+     * - rate limitting
+     * - logging context
+     * - tracing
+     */
+    private interceptService<T extends ServiceType>(type: T): ProxyHandler<ServiceImpl<T>> {
+        const grpc_service = type.typeName;
+        const self = this;
+        return {
+            get(target, prop) {
+                return (...args: any[]) => {
+                    const method = type.methods[prop as any];
+                    if (!method) {
+                        // Increment metrics for unknown method attempts
+                        console.warn("public api: unknown method", grpc_service, prop);
+                        const code = Code.Unimplemented;
+                        grpcServerStarted.labels(grpc_service, "unknown", "unknown").inc();
+                        grpcServerHandled.labels(grpc_service, "unknown", "unknown", Code[code]).inc();
+                        grpcServerHandling.labels(grpc_service, "unknown", "unknown", Code[code]).observe(0);
+                        throw new ConnectError("unimplemented", code);
+                    }
+                    const grpc_method = method.name;
+                    let grpc_type = "unknown";
+                    if (method.kind === MethodKind.Unary) {
+                        grpc_type = "unary";
+                    } else if (method.kind === MethodKind.ServerStreaming) {
+                        grpc_type = "server_stream";
+                    } else if (method.kind === MethodKind.ClientStreaming) {
+                        grpc_type = "client_stream";
+                    } else if (method.kind === MethodKind.BiDiStreaming) {
+                        grpc_type = "bidi_stream";
+                    }
+
+                    grpcServerStarted.labels(grpc_service, grpc_method, grpc_type).inc();
+                    const stopTimer = grpcServerHandling.startTimer({ grpc_service, grpc_method, grpc_type });
+                    const deferred = new Deferred<ConnectError | undefined>();
+                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                    deferred.promise.then((err) => {
+                        const grpc_code = err ? Code[err.code] : "OK";
+                        grpcServerHandled.labels(grpc_service, grpc_method, grpc_type, grpc_code).inc();
+                        stopTimer({ grpc_code });
+                    });
+
+                    const context = args[1] as HandlerContext;
+                    async function call<T>(): Promise<T> {
+                        const user = await self.verify(context);
```

Comment:
nit: As discussed IMO we could pull this (and some of the metrics here) into middleware. But not a priority for this PR.